src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -235,64 +299,72 @@ export class SavedObjectsService
       ),
     }));
 
-    const createSORepository = (callCluster: APICaller, extraTypes: string[] = []) => {
+    /**
+     * Note: We want to ensure that migrations have completed before
+     * continuing with further Core startup steps that might use SavedObjects
+     * such as running the legacy server, legacy plugins and allowing incoming
+     * HTTP requests.
+     *
+     * However, our build system optimize step and some tests depend on the
+     * HTTP server running without an Elasticsearch server being available.
+     * So, when the `migrations.skip` is true, we skip migrations altogether.
+     */
+    const cliArgs = this.coreContext.env.cliArgs;
+    const skipMigrations = cliArgs.optimize || savedObjectsConfig.skip;
+
+    this.createRepository = (callCluster: APICaller, extraTypes: string[] = []) => {
       return SavedObjectsRepository.createRepository(
-        migrator,
-        savedObjectSchemas,
-        setupDeps.legacyPlugins.pluginExtendedConfig,
+        this.migrator!,
+        new SavedObjectsSchema(this.schemas),
+        this.setupDeps!.legacyPlugins.pluginExtendedConfig,
         kibanaConfig.index,
         callCluster,
         extraTypes
       );
     };
```

Comment:
I'm creating the `createRepository` closure in migration method because all required properties are already there. this avoid duplicating +- 10 lines in `start`